paging/common/src/main/kotlin/androidx/paging/PageEvent.kt

```diff
@@ -216,6 +199,16 @@ internal sealed class PageEvent<T : Any> {
              */
             internal fun canDispatchWithoutInsert(loadState: LoadState, fromMediator: Boolean) =
                 loadState is LoadState.Loading || loadState is LoadState.Error || fromMediator
+
+            internal fun canDispatchWithoutInsert(combinedLoadStates: CombinedLoadStates): Boolean {
```

Comment:
`PageFetcher` and `PageFetcherSnapshot` already seem to be using the correct version of the function, as they have the previous state already. Would it by chance make more sense to do these checks in `CachedPageEventFlow`, as that's the only place this incorrect logic was used?

Reply:
This check is to ensure we only update to `LoadState.NotLoading `on Insert events, except in the case of Remote events.

I meant that the previous state is not passed in here, so it's impossible to tell what changed (if a loadType changed to NotLoading). For example, `PageFetcher` is still using this in `dispatchIfValid`.

In order to make this check do the right thing we need to pass in the previous state.

Reply:
Ahh okay I see, I think that makes a bit more sense, thank you for clarifying! Just one more question on this quick though:

In the case of the `PageFetcher`'s `dispatchIfValid` function for example, would it still need to pass in the previous state? Since `dispatchIfValid` is only being called from the coroutine right below it (here: https://github.com/androidx/androidx/pull/130/files#diff-4a907aa603257fb93c262b5dd0eab145ff03f06387f7732fc9f77619331a0a2bL145), does it not already perform the required checks before it's called? It seems to be checking if the state has changed first, and if so, calling `dispatchIfValid` with the new state. So in `dispatchIfValid` if we check if the state is `LoadState.NotLoading`, wouldn't it be guaranteed for it to have just changed to that state for any loadType? Additionally, since they're all remote event updates, wouldn't they all be valid to dispatch here regardless? Or am I misunderstanding how that works here?

Sorry for the excessive questions on this one, just trying to make sure I understand the logic here properly :sweat_smile:

Reply:
I just realized you also kept the old one - the code you linked above just calls into the override above, so that code path isn't affected. Although that isn't the only place this check is used. It's also used in `CachedPageEventFlow`. Either way, it would be better to make it robust rather than depend on its current use-sites because the latter will change over time.

Btw, instead of sending 3 updates there, you should just send one and call into this helper.

No worries on asking too many questions - it's important to get this bit right.

Regardless of the correctness of the use-sites though, this function not doing the correct thing seems mostly like a source of bugs down the road. Previously we were checking only what changed, now we are checking if any source states are `NotLoading`. I'm actually surprised the tests pass here so we probably should add some tests for this.

We're able to dispatch a load event dependent on what changed - so even if `source.refresh` is `Loading`, if it was already previously `Loading` that shouldn't prevent us from dispatching load state updates, because we might be dispatching the update to inform remote state, which is fine to do as a `LoadStateUpdate`.

The purpose of this, is so that source events is kept synchronous with `Insert` `PageEvents`, which means that code that reacts to `LoadState` can depend on the UI showing the loaded items.

Reply:
Yup, I had also noticed and fixed the use of it in `CachedPageEventFlow`. That all makes sense, I just made those changes now! Let me know if the logic seems correct this time, and I'll start creating tests around it.

Thank you again for going more in-depth on how that should work, I appreciate it :)